onnxruntime/core/optimizer/constant_folding.cc

```diff
@@ -182,9 +182,16 @@ Status ConstantFolding::ApplyImpl(Graph& graph, bool& modified, int graph_level,
 
     if (converted_to_constant) {
       // Remove single-output node chain for inputs of the node
-      for (auto p_ip_node = node->InputNodesBegin(); p_ip_node != node->InputNodesEnd(); ++p_ip_node) {
-        graph_utils::RemoveNodesWithOneOutputBottomUp(graph, *p_ip_node);
+      auto p_ip_node = node->InputNodesBegin();
+      const auto p_ip_node_end = node->InputNodesEnd();
+      while (p_ip_node != p_ip_node_end) {
+        const auto& input_node = *p_ip_node;
+        // Update the node iterator before removing the corresponding node because removing
+        // the node will invalidate the node iterator
+        ++p_ip_node;
```

Comment:
This logic is safe because the underlying STL container is a std::set. Other itrerators are not invalidated when an element is removed, only the iterator to the erased element is invalidated. Hence, this logic is safe.

See here - https://en.cppreference.com/w/cpp/container/set/erase
"References and iterators to the erased elements are invalidated. Other references and iterators are not affected."